File: apps/admin/src/api/types/search-reach.ts
import type { ScopeRef } from './scope-ref'

/**
 * How far a search looks (D30). Modelled as a union rather than three optional
 * fields so the impossible reaches cannot be written down: a collection with
 * no scope, or a scope with only an environment. The server takes the reach
 * from the URL path for the same reason — a forgotten query parameter would
 * *widen* a search rather than narrow it.
 */
export type SearchReach =
  | { kind: 'instance' }
  | { kind: 'scope'; scope: ScopeRef }
  | { kind: 'collection'; scope: ScopeRef; collection: string }

File: packages/silo-client-dart/lib/src/search/search_reach.dart
import '../transport/api_path.dart';

/// How far a search reaches. It is the receiver, never an argument.
final class SearchReach {
  SearchReach.collection(String project, String environment, String name)
    : path = ApiPath.collectionSearch(project, environment, name);

  SearchReach.environment(String project, String environment) : path = ApiPath.environmentSearch(project, environment);

  SearchReach.instance() : path = ApiPath.instanceSearch();

  final String path;
}
